ister.py

```diff
@@ -66,6 +66,90 @@
 
 LOG = None
 
+def extract_full_lines(text):
+    """Extract full lines from string 'text'. Return a tuple containing 2 elements
+    - the full lines and the last partial line.
+    """
+
+    full, partial = ("", "")
+    for line_match in re.finditer("(.*\n)|(.+$)", text):
+        if line_match.group(2):
+            partial = line_match.group(2)
+            break
+        full += line_match.group(1)
+    return (full, partial)
+
+def stream_fetcher(info, streamid):
+    """This function runs in a separate thread and fetches data from 'stream',
+    which is a file associated with the stdout or stderr pipes of a process.
+    """
+
+    partial = ""
+    stream = info["streams"][streamid]
+    try:
+        # Set non-blocking mode for the stream. We are doing this because we
+        # want to regularly supply the consumers with the output data and never
+        # block for too long.
+        fno = stream.fileno()
+        fcntl.fcntl(fno, fcntl.F_SETFL,
+                    fcntl.fcntl(fno, fcntl.F_GETFL) | os.O_NONBLOCK)
+        decoder = codecs.getincrementaldecoder('utf8')(errors="surrogateescape")
+        while not info["die_now"]:
+            # Wait for someting to appear in the stream. Wait for longest 1
+            # second in order to ensure we exit on "die_now".
+            if not select.select([stream], [], [], 1)[0]:
+                continue
+            data = None
+            try:
+                data = stream.read(4096)
+            except OSError as err:
+                if err.errno == errno.EAGAIN:
+                    continue
+                raise
+            if not data:
+                break
+            data = decoder.decode(data)
+            if not data:
+                continue
+            data, partial = extract_full_lines(partial + data)
+            if data:
+                info["queue"].put((streamid, data.strip()))
+    except BaseException as err:
+        LOG.error(err)
+
+    if partial:
+        info["queue"].put((streamid, partial))
```

Comment:
I don't think it matters much but I'm wondering if we should strip partial even though it won't have the trailing newline.

Reply:
Hrm looking at your next commit you take out strip on data so I guess that answers my question. I'll raise it on that commit.

Reply:
I'll amend the patches to make them more logical. I first have the .strip() but then realized it is incorrect, because it will strip _all_ the newlines - more than it should.

Reply:
Initially I had one commit, but then I split it to make review easier, but made a mistake. I've fixed this now.